### PR TITLE
[Builtins] Inline `SatInt` stuff and costing monoids

### PR DESCRIPTION
**Pull request by effectfully**
Just something I noticed in the Core, needs benchmarking, so don't look here yet.

### CODE REVIEW

File: plutus-core/plutus-core/src/Data/SatInt.hs [hunk 1]
```diff
@@ -49,49 +49,83 @@ instance Read SatInt where
...
 instance Enum SatInt where
-  succ (SI x) = SI (succ x)
-  pred (SI x) = SI (pred x)
-  toEnum                = SI
-  fromEnum              = unSatInt
+  {-# INLINE succ #-}
+  succ = coerce (succ :: Int -> Int)
+
+  {-# INLINE pred #-}
+  pred = coerce (pred :: Int -> Int)
```
Comment (effectfully) [inline plutus-core/plutus-core/src/Data/SatInt.hs:122]: @michaelpj we still have this bug.

File: plutus-core/plutus-core/src/Data/SatInt.hs [hunk 2]
```diff
@@ -49,49 +49,83 @@ instance Read SatInt where
...
     | y == minBound = maxBound
     | otherwise     = SI (negate y)
+
+  {-# INLINE abs #-}
   abs x
-    | x >= 0        = x
-    | otherwise     = negate x
-  signum (SI x)     = SI (signum x)
+    | x >= 0    = x
+    | otherwise = negate x
```
Comment (effectfully) [inline plutus-core/plutus-core/src/Data/SatInt.hs:91]: Why is it a custom definition? How does it handle near-maximum values? Needs a few tests at the very least.

Comment (michaelpj) [reply plutus-core/plutus-core/src/Data/SatInt.hs:91]: Basically this stuff all got copied from `safeint`... which sadly means I don't have good justifications for a bunch of things :/

Comment (kwxm) [reply plutus-core/plutus-core/src/Data/SatInt.hs:91]: I remember thinking at the time that we should dump everything from `safeInt` except for the few things we actually need.

**Comment by kwxm** [comment]
^ Just experimenting to see if running the benchmarks here uses the same machine that we use manually.  Looks like they do.  I've cancelled the benchmarking job now.

File: plutus-core/plutus-core/src/Data/SatInt.hs (hunk 2 above)
Comment (effectfully) [reply plutus-core/plutus-core/src/Data/SatInt.hs]: I've checked it and I think the current behavior is as correct as the default one:

```
>>> (minBound :: SatInt, negate (minBound :: SatInt))
(-9223372036854775808,9223372036854775807)
>>> (minBound :: SatInt, 0 - minBound :: SatInt)
(-9223372036854775808,9223372036854775807)
 
>>> (maxBound :: SatInt, negate (maxBound :: SatInt))
(9223372036854775807,-9223372036854775807)
>>> (maxBound :: SatInt, 0 - maxBound :: SatInt)
(9223372036854775807,-9223372036854775807)
```

Maybe it's a custom definition to make it faster or clearer.

Would still be great to have such tests if they're not there.

**Review by michaelpj — APPROVED** [review]
Looks fine!

**Pull request closed without merging**